crates/doc/src/inference.rs

```diff
@@ -1484,6 +1561,152 @@ impl Shape {
+
+                let partial_stringshape = StringShape {
+                    content_encoding: Some("base64".to_string()),
+                    ..Default::default()
+                };
+
+                self.string = StringShape::union(self.string.clone(), partial_stringshape);
+                false
+            }
+            crate::Node::Null => {
+                self.type_ = self.type_ | types::NULL;
+                false
+            }
+            crate::Node::Number(_num) => {
+                self.type_ = self.type_ | types::INT_OR_FRAC;
+                false
+            }
+            crate::Node::String(_s) => {
+                self.type_ = self.type_ | types::STRING;
+                let partial_stringshape = StringShape {
+                    // TODO: implement format widening
+                    ..Default::default()
+                };
+
+                self.string = StringShape::union(self.string.clone(), partial_stringshape);
+
+                false
+            }
+        }
+    }
+
+    // Prune any locations in this shape that have more than the allowed fields,
+    // squashing those fields into the `additionalProperties` subschema for that location.
+    pub fn enforce_field_count_limits(&mut self, loc: Location) {
+        // TODO: If we implement inference/widening of array tuple shapes
+        // then we'll need to also check that those aren't excessively large.
+        if !self.type_.overlaps(types::OBJECT) {
+            return;
+        }
+
+        let limit = match loc {
+            Location::Root => MAX_ROOT_FIELDS,
+            _ => MAX_NESTED_FIELDS,
+        };
+
+        if self.object.properties.len() > limit {
+            // Take all of the properties' shapes and
+            // union them into additionalProperties
+
+            let existing_additional_properties = self
+                .object
+                .additional
+                // `Shape::union` takes owned Shapes which is why we
+                // have to take ownership here.
+                .take()
+                .map(|boxed| *boxed)
+                .unwrap_or(Shape {
+                    // Start out maximally constrained
+                    type_: types::INVALID,
+                    provenance: Provenance::Inline,
+                    ..Default::default()
+                });
+
+            let merged_additional_properties = self
+                .object
+                .properties
+                // As part of squashing all known property shapes together into
+                // additionalProperties, we need to also remove those explicit properties.
+                .drain(..)
+                .fold(existing_additional_properties, |accum, prop| {
+                    Shape::union(accum, prop.shape)
+                });
+
+            self.object.additional = Some(Box::new(merged_additional_properties));
+        } else {
+            // We don't need to recur above because we already wiped out all
```

Comment:
yes we do. i could have a lot of properties which are all objects, also having a lot of properties.

Reply:
Neat, didn't think of this. Added a failing test and fixed it 👍